File: backend/src/models/user_loop.py
"""Data models for user loop/stem ingestion."""
from dataclasses import dataclass, field
from typing import Optional, List
import uuid

# Valid role assignments for user loop stems
VALID_ROLES = [
    "drums", "bass", "chord", "lead", "vocal",
    "fx", "percussion", "texture",
]


@dataclass
class UserLoopStem:
    """Represents a single user-uploaded loop stem."""
    id: str
    filename: str
    role: str  # one of VALID_ROLES
    loop_length_bars: int
    start_bar: float  # trim start (within the loop)
    end_bar: float  # trim end (within the loop)
    audio_path: str
    duration_seconds: float = 0.0

    def __post_init__(self):
        if not self.id:
            self.id = str(uuid.uuid4())
        if self.role not in VALID_ROLES:
            raise ValueError(
                f"Invalid role '{self.role}'. Must be one of: {VALID_ROLES}"
            )

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "filename": self.filename,
            "role": self.role,
            "loopLengthBars": self.loop_length_bars,
            "startBar": self.start_bar,
            "endBar": self.end_bar,
            "audioPath": self.audio_path,
            "durationSeconds": self.duration_seconds,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "UserLoopStem":
        return cls(
            id=data.get("id", str(uuid.uuid4())),
            filename=data["filename"],
            role=data["role"],
            loop_length_bars=data["loopLengthBars"],
            start_bar=data.get("startBar", 0.0),
            end_bar=data.get("endBar", float(data["loopLengthBars"])),
            audio_path=data.get("audioPath", ""),
            duration_seconds=data.get("durationSeconds", 0.0),
        )


@dataclass
class UserLoopBundle:
    """A collection of user loop stems for a project."""
    id: str
    project_id: str  # reference ID this is associated with
    stems: List[UserLoopStem] = field(default_factory=list)
    bpm: float = 0.0
    detected_bpm: float = 0.0
    bpm_ratio: float = 1.0  # ratio for time-stretching to match reference

    def __post_init__(self):
        if not self.id:
            self.id = str(uuid.uuid4())

    def compute_bpm_ratio(self, reference_bpm: float) -> float:
        """Compute the time-stretch ratio needed to match reference BPM.

        Returns:
            Ratio to multiply loop duration by. >1 means slow down, <1 means speed up.
        """
        if self.bpm <= 0 or reference_bpm <= 0:
            return 1.0
        self.bpm_ratio = reference_bpm / self.bpm
        return self.bpm_ratio

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "projectId": self.project_id,
            "stems": [s.to_dict() for s in self.stems],
            "bpm": self.bpm,
            "detectedBpm": self.detected_bpm,
            "bpmRatio": self.bpm_ratio,
        }

File: frontend/src/api/userLoop.ts
/**
 * API client for user loop/stem endpoints (Phase 4).
 */

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000';

/**
 * Valid role assignments for user loop stems.
 */
export const VALID_ROLES = [
  'drums', 'bass', 'chord', 'lead', 'vocal',
  'fx', 'percussion', 'texture',
] as const;

export type StemRole = typeof VALID_ROLES[number];

/**
 * Display labels for roles.
 */
export const ROLE_LABELS: Record<StemRole, string> = {
  drums: 'Drums',
  bass: 'Bass',
  chord: 'Chord / Harmony',
  lead: 'Lead / Melody',
  vocal: 'Vocal',
  fx: 'FX / Transition',
  percussion: 'Percussion',
  texture: 'Texture / Pad',
};

/**
 * Color mapping for roles.
 */
export const ROLE_COLORS: Record<StemRole, string> = {
  drums: '#EF5350',
  bass: '#42A5F5',
  chord: '#66BB6A',
  lead: '#FFA726',
  vocal: '#AB47BC',
  fx: '#26C6DA',
  percussion: '#EC407A',
  texture: '#78909C',
};

export interface UserLoopStem {
  id: string;
  filename: string;
  role: StemRole;
  loopLengthBars: number;
  startBar: number;
  endBar: number;
  audioPath: string;
  durationSeconds: number;
}

export interface UserLoopBundle {
  id: string;
  projectId: string;
  stems: UserLoopStem[];
  bpm: number;
  detectedBpm: number;
  bpmRatio: number;
}

export interface UploadLoopsResponse {
  projectId: string;
  bundleId: string;
  bpm: number;
  detectedBpm: number;
  bpmRatio: number;
  stemCount: number;
  newStems: UserLoopStem[];
  allStems: UserLoopStem[];
}

export interface GetLoopsResponse {
  projectId: string;
  bundle: UserLoopBundle | null;
  stemCount: number;
}

/**
 * Upload one or more loop stems for a project.
 */
export async function uploadUserLoops(
  projectId: string,
  files: File[],
  bpmOverride?: number,
): Promise<UploadLoopsResponse> {
  const formData = new FormData();
  files.forEach((file) => formData.append('files', file));

  let url = `${API_BASE_URL}/api/project/${projectId}/loops/upload`;
  if (bpmOverride !== undefined) {
    url += `?bpm_override=${bpmOverride}`;
  }

  const response = await fetch(url, {
    method: 'POST',
    body: formData,
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Upload failed' }));
    throw new Error(error.detail || `Upload failed with status ${response.status}`);
  }

  return response.json();
}

/**
 * Fetch all user loop stems for a project.
 */
export async function getUserLoops(projectId: string): Promise<GetLoopsResponse> {
  const response = await fetch(`${API_BASE_URL}/api/project/${projectId}/loops`);
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Failed to fetch loops' }));
    throw new Error(error.detail || `Failed with status ${response.status}`);
  }
  return response.json();
}

/**
 * Update the role assignment for a stem.
 */
export async function updateStemRole(
  projectId: string,
  stemId: string,
  role: StemRole,
): Promise<{ stemId: string; role: string; stem: UserLoopStem }> {
  const response = await fetch(
    `${API_BASE_URL}/api/project/${projectId}/loops/${stemId}/role`,
    {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ role }),
    },
  );
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Failed to update role' }));
    throw new Error(error.detail || `Failed with status ${response.status}`);
  }
  return response.json();
}

/**
 * Update loop length and trim boundaries for a stem.
 */
export async function updateStemBounds(
  projectId: string,
  stemId: string,
  bounds: { loopLengthBars?: number; startBar?: number; endBar?: number },
): Promise<{ stemId: string; stem: UserLoopStem }> {
  const response = await fetch(
    `${API_BASE_URL}/api/project/${projectId}/loops/${stemId}/bounds`,
    {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(bounds),
    },
  );
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Failed to update bounds' }));
    throw new Error(error.detail || `Failed with status ${response.status}`);
  }
  return response.json();
}

/**
 * Delete a user loop stem.
 */
export async function deleteUserStem(
  projectId: string,
  stemId: string,
): Promise<void> {
  const response = await fetch(
    `${API_BASE_URL}/api/project/${projectId}/loops/${stemId}`,
    { method: 'DELETE' },
  );
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Failed to delete stem' }));
    throw new Error(error.detail || `Failed with status ${response.status}`);
  }
}

/**
 * Update the BPM for the user's loop bundle.
 */
export async function updateLoopBpm(
  projectId: string,
  bpm: number,
): Promise<{ projectId: string; bpm: number; bpmRatio: number; bundle: UserLoopBundle }> {
  const response = await fetch(
    `${API_BASE_URL}/api/project/${projectId}/loops/bpm`,
    {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ bpm }),
    },
  );
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Failed to update BPM' }));
    throw new Error(error.detail || `Failed with status ${response.status}`);
  }
  return response.json();
}
